graspy/inference/semipar.py

```diff
@@ -0,0 +1,200 @@
+                normX1 = np.sum(X1**2, axis=1)
+                normX2 = np.sum(X2**2, axis=1)
+                normX1[normX1 <= 1e-15] = 1
+                normX2[normX2 <= 1e-15] = 1
+                X1 = X1 / np.sqrt(normX1[:, None])
+                X2 = X2 / np.sqrt(normX2[:, None])
+                R = orthogonal_procrustes(X1, X2)[0]
+                return np.linalg.norm(X1 @ R - X2)
+        else:
+            # in the omni case we don't need to align
+            return np.linalg.norm(X1 - X2)
+
+    def _embed(self, A1, A2):
+        if self.embedding == 'ase':
+            X1_hat = AdjacencySpectralEmbed(
+                n_components=self.n_components).fit_transform(A1)
+            X2_hat = AdjacencySpectralEmbed(
+                n_components=self.n_components).fit_transform(A2)
+        elif self.embedding == 'omnibus':
+            X_hat_compound = OmnibusEmbed(
+                n_components=self.n_components).fit_transform((A1, A2))
+            X1_hat = X_hat_compound[:A1.shape[0], :]
+            X2_hat = X_hat_compound[A2.shape[0]:, :]
+        return (X1_hat, X2_hat)
+
+    def fit(self, A1, A2):
+        """
+        Fits the test to the two input graphs
+
+        Parameters
+        ----------
+        A1, A2 : nx.Graph, nx.DiGraph, nx.MultiDiGraph, nx.MultiGraph, np.ndarray
+            The two graphs to run a hypothesis test on.
+            If np.ndarray, shape must be ``(n_vertices, n_vertices)`` for both graphs, 
+            where ``n_vertices`` is the same for both
+        
+        Returns
+        -------
+        p : float
+            The p value corresponding to the specified hypothesis test
+        """
+        A1 = import_graph(A1)
+        A2 = import_graph(A2)
+        if not is_symmetric(A1) or not is_symmetric(A2):
+            raise NotImplementedError()  # TODO asymmetric case
+        if A1.shape != A2.shape:
+            raise ValueError('Input matrices do not have matching dimensions')
+        if self.n_components is None:
+            # get the last elbow from ZG for each and take the maximum
+            num_dims1 = select_dimension(A1)[0][-1]
+            num_dims2 = select_dimension(A2)[0][-1]
+            self.n_components = max(num_dims1, num_dims2)
+        X_hats = self._embed(A1, A2)
+        T_sample = self._difference_norm(X_hats[0], X_hats[1])
+        T1_bootstrap = self._bootstrap(X_hats[0])
+        T2_bootstrap = self._bootstrap(X_hats[1])
+
+        # Continuity correction - note that the +0.5 causes p > 1 sometimes # TODO
+        p1 = (len(T1_bootstrap[T1_bootstrap >= T_sample]) +
+              0.5) / self.n_bootstraps
+        p2 = (len(T2_bootstrap[T2_bootstrap >= T_sample]) +
+              0.5) / self.n_bootstraps
+
+        p = max(p1, p2)
+
+        # TODO : what to store as fields here, or make _private fields
+        # at least for the sake of testing, I'm going to keep everything
+
+        self.T1_bootstrap = T1_bootstrap
```

Comment:
If you are keeping these, make sure they end with `_`. The underscore tells that it's an attribute that is saved after some computation, not a parameter that was originally input.